src/main/java/io/cdap/plugin/http/source/batch/HttpRecordReader.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright © 2019 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package io.cdap.plugin.http.source.batch;
+
+import io.cdap.plugin.http.source.common.pagination.BaseHttpPaginationIterator;
+import io.cdap.plugin.http.source.common.pagination.PaginationIteratorFactory;
+import org.apache.commons.lang3.tuple.ImmutablePair;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.io.LongWritable;
+import org.apache.hadoop.io.Text;
+import org.apache.hadoop.mapreduce.InputSplit;
+import org.apache.hadoop.mapreduce.RecordReader;
+import org.apache.hadoop.mapreduce.TaskAttemptContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+
+/**
+ * RecordReader implementation, which reads text records representations and http codes
+ * using {@link BaseHttpPaginationIterator} subclasses.
+ */
+public class HttpRecordReader extends RecordReader<LongWritable, Text> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(HttpRecordReader.class);
+  private BaseHttpPaginationIterator httpPagionationIterator;
+  private LongWritable key;
+  private Text value;
+
+  /**
+   * Initialize an iterator and config.
+   *
+   * @param inputSplit specifies batch details
+   * @param taskAttemptContext task context
+   */
+  @Override
+  public void initialize(InputSplit inputSplit, TaskAttemptContext taskAttemptContext) {
+    Configuration conf = taskAttemptContext.getConfiguration();
+    HttpBatchSourceConfig httpBatchSourceConfig = HttpBatchSourceConfig.fromMapredureConfig(conf);
+    httpPagionationIterator = PaginationIteratorFactory.createInstance(httpBatchSourceConfig);
+  }
+
+  @Override
+  public boolean nextKeyValue() {
+    if (!httpPagionationIterator.hasNext()) {
+      return false;
+    }
+    ImmutablePair<Integer, String> pair = httpPagionationIterator.next();
+    key = new LongWritable(pair.getLeft());
+    value = new Text(pair.getRight());
+    return true;
+  }
+
+  @Override
+  public LongWritable getCurrentKey() {
+    return key;
+  }
+
+  @Override
+  public Text getCurrentValue() {
+    return value;
+  }
+
+  @Override
+  public float getProgress() {
+    // progress is unknown
+    return 1.0f;
```

Comment:
should probably use 0 instead of 1. Showing no progress is better than showing full progress.

Reply:
done